### ISSUE TITLE
Question about Travis iOS builds

### ISSUE THREAD

**Issue by caseymcc** [ruslo/hunter#719]
I am trying to fix the failed tests on the ceres-solver package NeroBurner has posted. Most of the failures are just related to glog not compiling for a particular platform (I have disabled them and will post a pull request for it later). I am trying to fix 2 items that are related to ceres directly. One has taken me down a rabbit hole and I need some help on where/how it should be fixed.

This issue is on the iOS build (https://travis-ci.org/caseymcc/hunter/jobs/214737167) and ceres is setup for building iOS so I didn't want to just disable the test. The error occurs because ceres' cmake is checking to verify that the iOS version is supported. The issue comes because ceres is expecting cmake to be ran with a toolchain file (https://github.com/caseymcc/ceres-solver/blob/hunter/cmake/iOS.cmake) before anything else (the toolchain file sets up some cmake vars that the main cmake file uses).

I have tracked as much as I could through hunter looking for a way to setup a toolchain file but nothing popped out at me considering the test builds already provide cmake with toolchain files. So I tracked all that back into polly and found that it has similar toolchain files as what ceres is using it just doesn't setup the same vars. In fact in the polly "os/iphone.cmake" file 

    # Emulate OpenCV toolchain --
    set(IOS YES)
    # -- end

IOS is the same flag ceres is expecting from its toolchain file to build for iOS. Ceres' toolchain file uses xocdebuild to fill out the vars:

    # Get the SDK version information.
    execute_process(COMMAND xcodebuild -sdk ${CMAKE_OSX_SYSROOT} -version SDKVersion
      OUTPUT_VARIABLE IOS_SDK_VERSION
      ERROR_QUIET
      OUTPUT_STRIP_TRAILING_WHITESPACE)

so it should work pretty good and the vars seemed to be properly named. So on to the questions;

1. Should polly adopt some of these variables from ceres and just ignore its own toolchain file
2. Find a way to call ceres' toolchain file
3. Or something else?

**Comment by caseymcc** [comment]
After thinking about it for a second or 2, changing polly will have little to no affect on everyone else's build environment so I am guessing I need to make sure ceres' toolchain file is called. Is there an existing way to do this?

**Comment by ruslo** [comment]
> Should polly adopt some of these variables from ceres and just ignore its own toolchain file

Variable `IOS` originally introduced to fix OpenCV build, that's true. However I don't think it is a good idea to add random variables from projects. `IOS` is straight and simple, I'm pretty sure it will be added to CMake when iOS will be officially supported. Can you give an example of missing variable?

> Find a way to call ceres' toolchain file

No, as far as I can see this toolchain file can deal with only for one architecture. Polly's toolchain support universal library build.

> Or something else?

Any specific problems you can post here? I think we can deal with it. With Polly's toolchain you don't have to modify CMakeLists.txt at all. If C++ code follow standard and CMake code written without tricks then library just works even if iOS support is not originally planned.

> -- Building Ceres for iOS platform:
> CMake Error at CMakeLists.txt:180 (message):
> Unsupported iOS version: , Ceres requires at least iOS version 7.0

What we are checking here and how?

**Comment by caseymcc** [comment]
as you noted above ceres is looking for 

    IOS_DEPLOYMENT_TARGET

which in the current build ends up being empty. The ceres cmake files are expecting cmake to be called like this

    cmake \
    -DCMAKE_TOOLCHAIN_FILE=../ceres-solver/cmake/iOS.cmake \
    -DEIGEN_INCLUDE_DIR=/path/to/eigen/header \
    -DIOS_PLATFORM=<PLATFORM> \
    <PATH_TO_CERES_SOURCE>

and when the iOS.cmake file runs it performs some data gathering that fills out the following cmake vars using calls to xcodebuild/ etc...

    XCODE_VERSION
    IOS_PLATFORM -if not set
    CMAKE_OSX_SYSROOT -if not set
    IOS_SDK_VERSION
    IOS_PLATFORM_SDK_DIR
    CMAKE_IOS_DEVELOPER_ROOT
    CMAKE_C_COMPILER -if not set
    CMAKE_CXX_COMPILER -if not set
    IOS_LIBTOOL
    CMAKE_HOST_SYSTEM_VERSION
    CMAKE_SYSTEM_NAME -to Darwin
    CMAKE_SYSTEM_VERSION -set to IOS_SDK_VERSION
    UNIX
    APPLE
    IOS
    CMAKE_OSX_DEPLOYMENT_TARGET -set to empty
    #sets compiler with some options
    IOS_DEPLOYMENT_TARGET -set to IOS_SDK_VERSION

and adds 2 macros (which don't seem to be used).

    set_xcode_property
    find_host_package

I am going to attempt to call the iOS.make file from the ceres' cmake setup. I can see there is already some issues that I will have to hunt down.

**Comment by ruslo** [comment]
> as you noted above ceres is looking for IOS_DEPLOYMENT_TARGET

It's `IOS_DEPLOYMENT_SDK_VERSION` in Polly:
* https://github.com/ruslo/polly/blob/7f57a8ece4dc2b19d472beb0e48fe2d199c99dda/ios-10-2-dep-9-3-armv7.cmake#L14-L15

> and when the iOS.cmake file runs it performs some data gathering that fills out the following cmake vars using calls to xcodebuild/ etc...

The more important question is what variables really used in ceres-solver.

My suggestion is to try to set `IOS_DEPLOYMENT_SDK_VERSION` instead of `IOS_DEPLOYMENT_TARGET` and see what error we will have next.

**Comment by alexsmac** [comment]
Hi, Ceres doesn't actually reference very much at all in its CMakeLists when building for iOS, specifically the only variables it requires are IOS and IOS_DEPLOYMENT_TARGET.  It requires the latter to verify that the iOS version to build for is >= 7.0, because pre-7.0 versions used libstdc++ (vs libc++ for >= 7.0) and doing so causes problems with the detection of shared_ptr & unordered_map.

The choice of IOS_DEPLOYMENT_TARGET is for consistency with the official CMake variable [CMAKE_OSX_DEPLOYMENT_TARGET](https://cmake.org/cmake/help/v3.8/variable/CMAKE_OSX_DEPLOYMENT_TARGET.html) which serves exactly the same purpose, albeit for OS X (specifying `-mmacosx-version-min` vs `-mios-version-min`).

**Comment by caseymcc** [comment]
Yeah I noticed that about the vars that were setup, however there were some compiler flags setup by the iOS.cmake. Since I didn't want to hunt them down what they meant, I just modified ceres' cmake files to include iOS.cmake if IOS was set (which is true from the polly setup, and would be needed for the average setup). I had to empty CMAKE_OSX_SYSROOT as its value coming from polly did not work. Emptying it allowed the iOS.cmake file to query it from xcode which seems to have worked just fine.

The only issue I have left at the moment is that in its current state it is building ceres for multiple architectures armv7, armv7s, arm64, x86, x86_64 which is overrunning Travis' build log limit and causing the build to show as failed. Just need to hunt down why all of the architectures are getting built.

**Comment by ruslo** [comment]
> Hi, Ceres doesn't actually reference very much at all in its CMakeLists when building for iOS, specifically the only variables it requires are IOS and IOS_DEPLOYMENT_TARGET.

@alexsmac Thanks for the info. @caseymcc It means no big problems expected.

> The choice of IOS_DEPLOYMENT_TARGET is for consistency with the official CMake variable CMAKE_OSX_DEPLOYMENT_TARGET which serves exactly the same purpose, albeit for OS X

@alexsmac Okay, I see. Good point.

> The only issue I have left at the moment is that in its current state it is building ceres for multiple architectures armv7, armv7s, arm64, x86, x86_64 which is overrunning Travis' build log limit and causing the build to show as failed. Just need to hunt down why all of the architectures are getting built.

@caseymcc It's the default:
* http://polly.readthedocs.io/en/latest/toolchains/ios.html#ios-x-y

@caseymcc You can use your set of architectures on Travis CI to reduce build time:
* http://polly.readthedocs.io/en/latest/toolchains/ios.html#ios-x-y-arch1-arch2

> I had to empty CMAKE_OSX_SYSROOT as its value coming from polly did not work. Emptying it allowed the iOS.cmake file to query it from xcode which seems to have worked just fine.

@caseymcc It means users will not be able to install simulator+devices universal libraries. With the all info above I see no problem with using Polly iOS toolchain instead of project-specific one. If the `IOS_DEPLOYMENT_TARGET` -> `IOS_DEPLOYMENT_SDK_VERSION` do the trick for you I'm ready to add this variable to Polly, meaning that you don't have to modify `ceres-solver/CMakeLists.txt` at all.

**Comment by caseymcc** [comment]
I hacked in a test to see if setting IOS_DEPLOYMENT_TARGET alone would work. Ceres compiled but the hunter example program was unable to compile. (https://travis-ci.org/caseymcc/hunter/builds/215546885). I used "env: PROJECT_DIR=examples/ceres-solver TOOLCHAIN=ios-nocodesign-9-3-armv7" for the test.

    /Users/travis/build/caseymcc/hunter/examples/ceres-solver/foo.cpp:1:10: fatal error: 'ceres/ceres.h' file not found

looking at the compiler call there is nothing including the hunter install directory,

    -I/Users/travis/build/caseymcc/hunter/_testing/_builds/ios-nocodesign-9-3-armv7/Release-iphoneos/include
    -I/Users/travis/build/caseymcc/hunter/_testing/_builds/ios-nocodesign-9-3-armv7/download-ceres-solver.build/Release-iphoneos/foo.build/DerivedSources/armv7 
    -I/Users/travis/build/caseymcc/hunter/_testing/_builds/ios-nocodesign-9-3-armv7/download-ceres-solver.build/Release-iphoneos/foo.build/DerivedSources

were as on the other builds, (osx, linux, windows, etc..) you can see the hunter install inclusion.

    -I/Users/travis/build/caseymcc/hunter/_testing/_builds/osx-10-11/Release/include
    -I/Users/travis/build/caseymcc/hunter/_testing/Hunter/_Base/672c979/77607d5/cdb2970/Install/include
    -I/Users/travis/build/caseymcc/hunter/_testing/Hunter/_Base/672c979/77607d5/cdb2970/Install/include/eigen3 
    -I/Users/travis/build/caseymcc/hunter/_testing/_builds/osx-10-11/download-ceres-solver.build/Release/foo.build/DerivedSources/x86_64
    -I/Users/travis/build/caseymcc/hunter/_testing/_builds/osx-10-11/download-ceres-solver.build/Release/foo.build/DerivedSources -Wmost

ceres is found by the example projects find_package

    -- [hunter *** DEBUG *** 2017-03-27T14:28:38] load: /Users/travis/build/caseymcc/hunter/_testing/Hunter/_Base/Download/Hunter/unknown/f2d44a0/Unpacked/cmake/projects/ceres-solver/hunter.cmake ... end
    -- Found Ceres compiled with miniglog substitute for glog, beware this will likely cause problems if glog is later linked.
    -- Found required Ceres dependency: Eigen version 3.3.1 in /Users/travis/build/caseymcc/hunter/_testing/Hunter/_Base/f2d44a0/77607d5/52cf9f8/Install/include/eigen3
    -- Found Ceres version: 1.12.0 installed in: /Users/travis/build/caseymcc/hunter/_testing/Hunter/_Base/f2d44a0/77607d5/52cf9f8/Install with components: [SchurSpecializations]

I checked up further in the iOS log to verify ceres installed:

    -- Install configuration: "Release"
    -- Installing: /Users/travis/build/caseymcc/hunter/_testing/Hunter/_Base/f2d44a0/77607d5/52cf9f8/Install/include/ceres/autodiff_cost_function.h
    -- Installing: /Users/travis/build/caseymcc/hunter/_testing/Hunter/_Base/f2d44a0/77607d5/52cf9f8/Install/include/ceres/autodiff_local_parameterization.h
    -- Installing: /Users/travis/build/caseymcc/hunter/_testing/Hunter/_Base/f2d44a0/77607d5/52cf9f8/Install/include/ceres/c_api.h
    -- Installing: /Users/travis/build/caseymcc/hunter/_testing/Hunter/_Base/f2d44a0/77607d5/52cf9f8/Install/include/ceres/ceres.h
    -- Installing: /Users/travis/build/caseymcc/hunter/_testing/Hunter/_Base/f2d44a0/77607d5/52cf9f8/Install/include/ceres/conditioned_cost_function.h
    etc...

any ideas, I don't have an iOS development environment to test with locally. (guess I might need to set one up)

**Comment by ruslo** [comment]
> looking at the compiler call there is nothing including the hunter install directory

Sounds like ceres-solver/CMakeLists.txt issue. This path should be a property of imported installed ceres-solver library.

**Comment by caseymcc** [comment]
I just found my issue. Neroburner had checked in his changes under hunter-1.12.0 and they were merged under hunter-packages/ceres-solver as the same branch name. I pulled hunter branch from hunter-packages/ceres-solver and have been pulling my hair out ever since wondering why Nero's worked and my has all kinds of issues (on all platforms). I have now pulled the right branch and will run some test in a few.
